packages/base-cli/src/index.ts

```diff
@@ -0,0 +1,212 @@
+import fs from 'fs';
+import path from 'path';
+import { parseArguments, SignalValue } from './args';
+import postcss from 'postcss';
+import type { PluginCreator } from 'postcss';
+
+export * from './help';
+
+type PluginCreatorOptions = Record<string, unknown> | null;
+
+export function cli(plugin: PluginCreator<PluginCreatorOptions>, allowedPluginOpts: Array<string>, helpLogger: () => void) {
+	// get process and plugin options from the command line
+	const argo = parseArguments(process.argv.slice(2), allowedPluginOpts, helpLogger);
+	if (argo === SignalValue.InvalidArguments) {
+		process.exit(1);
+	}
+
+	if (argo.stdin && argo.stdout) {
+		return getStdin().then((css) => {
+			if (!css) {
+				helpLogger();
+				process.exit(1);
+			}
+
+			const result = postcss([plugin(argo.pluginOptions)]).process(css, {
+				from: 'stdin',
+				to: 'stdout',
+				map: argo.inlineMap ? { inline: true } : false,
+			});
+
+			return result.css;
+		}).then((result) => {
+			process.stdout.write(result + (argo.inlineMap ? '\n' : ''));
+
+			process.exit(0);
+		}).catch((error) => {
+			console.error(argo.debug ? error : error.message);
+
+			process.exit(1);
+		});
```

Comment:
I'm not strongly against promises but I reckon that given we already return a promise, it'd be more legible if we go for `async`/`await`?

Reply:
`async`/`await` would be much better!
I haven't focussed on this aspect and got caught up in the argument parsing.

Thank you for mentioning it!

Reply:
Switched too `async/await` and removed some code which has nicer native API's.
(the starting point of this was old CLI code from one of the plugins)

Going to give it another look tomorrow to remove some duplication of split it into smaller functions.

Reply:
As answered in the other comment I chose to split first.
I think this is ready for another view :)